zaza/openstack/charm_tests/rabbitmq_server/tests.py

```diff
@@ -0,0 +1,406 @@
+# Copyright 2019 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""RabbitMQ Testing."""
+
+import json
+import logging
+import time
+import uuid
+
+import juju
+import zaza.model
+import zaza.openstack.charm_tests.test_utils as test_utils
+import zaza.openstack.utilities.generic as generic_utils
+
+from charmhelpers.core.host import (
+    lsb_release,
+    CompareHostReleases,
+)
+
+from . import utils as rmq_utils
+
+
+class RmqTests(test_utils.OpenStackBaseTest):
+    """Zaza tests on a basic rabbitmq cluster deployment."""
+
+    @classmethod
+    def setUpClass(cls):
+        """Run class setup for running tests."""
+        super(RmqTests, cls).setUpClass()
+
+    def _get_uuid_epoch_stamp(self):
+        """Return a string based on uuid4 and epoch time.
+
+        Useful in generating test messages which need to be unique-ish.
+        """
+        return '[{}-{}]'.format(uuid.uuid4(), time.time())
+
+    def _test_rmq_amqp_messages_all_units(self, units,
+                                          ssl=False, port=None):
+        """Reusable test to send/check amqp messages to every listed rmq unit.
+
+        Reusable test to send amqp messages to every listed rmq
+        unit. Checks every listed rmq unit for messages.
+        :param units: list of units
+        :returns: None if successful.  Raise on error.
+
+        """
+        # Add test user if it does not already exist
+        rmq_utils.add_user(units)
+
+        # Handle ssl (includes wait-for-cluster)
+        if ssl:
+            rmq_utils.configure_ssl_on(units, port=port)
+        else:
+            rmq_utils.configure_ssl_off(units)
+
+        # Publish and get amqp messages in all possible unit combinations.
+        # Qty of checks == (qty of units) ^ 2
```

Comment:
Non-blocking, but I would still like to get away from n^2. That is overkill. Because it is not only n^2 it is num_tests * n^2.